docs/async-await.md

```diff
@@ -0,0 +1,174 @@
+}
+```
+
+The main difference to today's `Request.Body` type is the lack of a `StreamWriter` for streaming scenarios. The existing StreamWriter offered the user an API to write into (thus the user was in control of when writing happened). The new `HTTPClientRequest.Body` uses `AsyncSequence`s to stream requests. By iterating over the provided AsyncSequence, the HTTPClient is  in control when writes happen, and can ask for more data efficiently. 
+
+Using the `AsyncSequence` from the Swift standard library as our upload stream mechanism dramatically reduces the learning curve for new users.
+
+### New `HTTPClientResponse` type
+
+The `HTTPClientResponse` looks more similar to the existing [`HTTPClient.Response`][HTTPClient.Response] type. The biggest difference is again the `body` property, which is now an `AsyncSequence` of `ByteBuffer`s instead of a single optional `ByteBuffer?`. This will make every response on AsyncHTTPClient streaming by default. As with `HTTPClientRequest`, we dropped the namespacing on `HTTPClient` to allow easier discovery with autocompletion.
+
+```swift
+public struct HTTPClientResponse {
+  /// the used http version
+  public var version: HTTPVersion
+  /// the http response status
+  public var status: HTTPResponseStatus
+  /// the response headers
+  public var headers: HTTPHeaders
+  /// the response payload as an AsyncSequence
+  public var body: Body
+}
+
+extension HTTPClientResponse {
+  public struct Body: AsyncSequence {
+    public typealias Element = ByteBuffer
+    public typealias AsyncIterator = Iterator
+  
+    public struct Iterator: AsyncIteratorProtocol {
+      public typealias Element = ByteBuffer
+    
+      public func next() async throws -> ByteBuffer?
+    }
+  
+    public func makeAsyncIterator() -> Iterator
+  }
+}
+```
+
+Note: The user must consume the `Body` stream or drop the `HTTPClientResponse`, to ensure that the 
+internal HTTPClient connection can move forward. Dropping the `HTTPClientResponse` would lead to a 
+request cancellation which in turn would lead to a close of an exisiting HTTP/1.1 connection.
+
+At a later point we could add trailers to the `HTTPClientResponse` as effectful properties:
+
+```swift
+    public var trailers: HTTPHeaders? { async throws }
+```
+
+However we will need to make sure that the user has consumed the body stream completely before, calling the trailers, because otherwise we might run into a situation from which we can not progress forward:
+
+```swift
+do {
+  var request = HTTPClientRequest(url: "https://swift.org/")
+  let response = try await httpClient.execute(request, deadline: .now() + .seconds(3))
+  
+  var trailers = try await response.trailers // can not move forward since body must be consumed before.
+} catch {
+  print(error)
+}
+```
+
+In such a case we can either throw an error or crash.
+
+### New invocation
+
+The new way to invoke a request shall look like this:
```

Comment:
Not exactly related the proposal, but will convenience methods be added similar to the existing api (e.g. `client.get(...)` etc.)?

Reply:
Added a FAQ to answer this question.